File: tests/managed_integration/fixtures/vbnet/Classes.vb
' Tier 2: Classes, inheritance, and interfaces. Tests VB.NET class hierarchy,
' overriding methods, and interface implementation.
Imports System.Collections.Generic

Interface IAnimal
    Function Speak() As String
    ReadOnly Property Name As String
End Interface

Class Animal
    Implements IAnimal
    Private _name As String
    Public Sub New(name As String)
        _name = name
    End Sub
    Public ReadOnly Property Name As String Implements IAnimal.Name
        Get
            Return _name
        End Get
    End Property
    Public Overridable Function Speak() As String Implements IAnimal.Speak
        Return "..."
    End Function
End Class

Class Dog
    Inherits Animal
    Public Sub New(name As String)
        MyBase.New(name)
    End Sub
    Public Overrides Function Speak() As String
        Return "Woof!"
    End Function
End Class

Class Cat
    Inherits Animal
    Public Sub New(name As String)
        MyBase.New(name)
    End Sub
    Public Overrides Function Speak() As String
        Return "Meow!"
    End Function
End Class

Module Classes
    Sub Main()
        Dim animals As New List(Of IAnimal) From {
            New Dog("Rex"),
            New Cat("Whiskers"),
            New Dog("Buddy")
        }
        For Each a As IAnimal In animals
            Console.WriteLine(a.Name & " says: " & a.Speak())
        Next
    End Sub
End Module

File: tests/managed_integration/fixtures/python/classes.py
# Tier 2: Classes and inheritance. Tests __init__, __repr__, __str__,
# property decorators, classmethods, and MRO.
class Animal:
    _registry = {}

    def __init__(self, name: str, sound: str):
        self.name = name
        self.sound = sound

    def __repr__(self) -> str:
        return f"{type(self).__name__}({self.name!r})"

    def speak(self) -> str:
        return f"{self.name} says {self.sound}"

    @classmethod
    def register(cls, subclass):
        cls._registry[subclass.__name__] = subclass
        return subclass

@Animal.register
class Dog(Animal):
    def __init__(self, name: str):
        super().__init__(name, "Woof")

    @property
    def is_good_boy(self) -> bool:
        return True

@Animal.register
class Cat(Animal):
    def __init__(self, name: str):
        super().__init__(name, "Meow")

if __name__ == "__main__":
    animals = [Dog("Rex"), Cat("Whiskers"), Dog("Buddy")]
    for a in animals:
        print(a.speak())
    print(repr(animals[0]))
    print("good_boy:", animals[0].is_good_boy)
    print("registry:", list(Animal._registry.keys()))
